ads/google/a4a/utils.js

```diff
@@ -194,17 +195,29 @@ export function googleBlockParameters(a4a, opt_experimentIds) {
  * @return {!Promise<!Object<string,!Array<!Promise<!../../../src/base-element.BaseElement>>>>}
  */
 export function groupAmpAdsByType(win, type, groupFn) {
+  // Look for amp-ad elements of correct type or those contained within
+  // standard container type.  Note that display none containers will not be
+  // included as they will never be measured.
+  // TODO(keithwrightbos): what about slots that become measured due to removal
+  // of display none (e.g. user resizes viewport and media selector makes
+  // visible).
   return Services.resourcesForDoc(win.document).getMeasuredResources(win,
-      r => r.element.tagName == 'AMP-AD' &&
-        r.element.getAttribute('type') == type)
-      .then(resources => {
-        const result = {};
-        resources.forEach(r => {
-          const groupId = groupFn(r.element);
-          (result[groupId] || (result[groupId] = [])).push(r.element.getImpl());
-        });
+      r => (Object.keys(ValidAdContainerTypes).includes(r.element.tagName) &&
+          !!r.element.querySelector(`amp-ad[type=${type}]`)) ||
+        (r.element.tagName == 'AMP-AD' &&
+          r.element.getAttribute('type') == type))
+      // Need to wait on any contained element resolution followed by build
+      // of child ad.
+      .then(resources => Promise.all(resources.map(
+          resource => resource.element.tagName == 'AMP-AD' ? resource.element :
+            whenUpgradedToCustomElement(dev().assertElement(
```

Comment:
I think we also need to wait for the element's whenBuilt() promise here. But actually I don't when when is sraRequest sent. can it be sent before buildCallback? If so, then we are good.

Reply:
We block sending the ad request on getAdUrl execution.  Given that happens after the instance is built, we should be fine as is.